syntax = "proto2";
package txnfs.proto;

enum TransactionType {
	NONE = 0;
	VCREATE = 1;
	VMKDIR = 2;
	VWRITE = 3;
	VRENAME = 4;
	VUNLINK = 5;
	VSYMLINK = 6;
	VOPEN = 7;
}

enum FileType {
	FT_NONE = 0;
	FT_FILE = 1;
	FT_DIRECTORY = 2;
	FT_SYMLINK = 3;
	FT_HARDLINK = 4;
}

// A 128-bit file ID.
message FileID {
	required bytes id = 1;
	optional FileType type = 2;
}

message ObjectId {
	optional bytes uuid = 1;  // uuid_t
	optional FileType type = 2;
}

// A CreatedObject represents a file or directory to be created in a compound.
// When recovering from a partially executed compound that may have created
// some objects, the created objects do not have mapping entries in the
// database. This is because establishing mapping between UUID and FH is the
// last step of the transaction execution. Therefore, we need to locate the
// created files by following |base| and |path|. For example, |base| may be the
// parent directory whereas |path| is the name of the created file; also |base|
// may be the root directory whereas |path| is the absolute path file of the
// created file. |base| and |path| can be easily retrieved from the NFS
// compound and |base| will exist for sure for any valid compound.
//
// Note that it is possible that |path| pre-existed before the compound (e.g.,
// open a file with O_CREATE) and we should not delete it during the recovery.
// We can check this by looking into the database: if the FH of |path| has a
// mapping in the database, then we know it pre-existed before the compound was
// executed.
message CreatedObject {
	// The base directory of this object. Its type must be FT_DIRECTORY.
	optional ObjectId base = 1;
	// The path from the base directory to the created object.
	optional string path = 2;
	// The pre-allocated ID for the object to be created.
	optional ObjectId allocated_id = 3;
}

message VCreateTxn {
	repeated FileID created_files = 1;  // deprecated
	repeated CreatedObject objects = 2;
}

// deprecated. Use VCreateTxn.
message VMkdirTxn {
	repeated FileID created_dirs = 1;  // deprecated
	repeated CreatedObject dirs = 2;
}

message VWriteTxn {
	// Inside the backup directory of each transaction, the file will be
	// named "txn_XXXX" where XXXX is the fileid in hex.
	required string backup_dir_path = 1;
	// A VWrite may create files, so |created_files| track the files have
	// been created by the pending transaction. When undoing a VWrite that
	// created files, the files inside |created_files| should be removed.
	//
	// For VWrite that does not create files, |created_files| will be
	// empty.
	repeated FileID created_files = 2;

	repeated CreatedObject files = 3;
}

message VRenameTxn {
	message Rename {
		required string src_path = 1;
		required string dst_path = 2;
		required FileID src_fileid = 3;
		// Present only if destination file exists.
		optional FileID dst_fileid = 4;
		// The renamed file-system object is a directory or not?
		optional bool is_directory = 5;
	}
	repeated Rename renamed_objs = 1;
	optional string backup_dir = 2;
}

message VUnlinkTxn {
	message Unlink {
		required ObjectId parent_id = 1;
		required string name = 2;
	}
	required string backup_dir_path = 1;
	repeated Unlink unlinked_objs = 2;
}

message VSymlinkTxn {
	message Symlink {
		required string src_path = 1;
		required ObjectId parent_id = 2;
		required string name = 3;
	}
	repeated Symlink linked_objs = 1;
}

// A undo log of pending transactions. A TransactionLog should contain all
// information needed for us to undo a pending transaction when the system
// crashes. Upon successul commits and rollbacks of transactions, these logs
// should be cleared.
// 
// TransactionLogs should be stored into a database, so that they are available
// when we recover from a crash. To recover, we should iterate all
// TransactionLogs in the database and undo all the pending transactions
// according to information inside TransactionLogs.
message TransactionLog {
	required int64 id = 1;  // a unique transaction id

	required TransactionType type = 2;

	// Depending on the value of |type|, only one of the following field
	// will present.
	optional VCreateTxn creates = 3;
	optional VMkdirTxn mkdirs = 4;
	optional VWriteTxn writes = 5;
	optional VRenameTxn renames = 6;
	optional VUnlinkTxn unlinks = 7;
	optional VSymlinkTxn symlinks = 8;
}
